ddtrace/appsec/processor.py

```diff
@@ -0,0 +1,96 @@
+import errno
+import json
+import os.path
+import threading
+from typing import TYPE_CHECKING
+
+import attr
+
+import ddtrace
+from ddtrace.appsec._ddwaf import DDWaf
+from ddtrace.constants import MANUAL_KEEP_KEY
+from ddtrace.ext import SpanTypes
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.processor import SpanProcessor
+from ddtrace.utils.formats import get_env
+
+
+if TYPE_CHECKING:
+    from ddtrace import Span
+
+ROOT_DIR = os.path.dirname(os.path.abspath(__file__))
+DEFAULT_RULES = os.path.join(ROOT_DIR, "rules.json")
+
+log = get_logger(__name__)
+
+
+def get_rules():
+    return get_env("appsec", "rules", default=DEFAULT_RULES)
+
+
+@attr.s(eq=False)
+class AppSecSpanProcessor(SpanProcessor):
+
+    _lock = attr.ib(init=False, factory=threading.Lock, repr=False)
+
+    rules = attr.ib(type=str, factory=get_rules)
+    _ddwaf = attr.ib(type=DDWaf, default=None)
+
+    @property
+    def enabled(self):
+        return self._ddwaf is not None
+
+    def __attrs_post_init__(self):
+        # type: () -> None
+        if self._ddwaf is None:
+            try:
+                with open(self.rules, "r") as f:
+                    rules = json.load(f)
+            except EnvironmentError as err:
+                if err.errno == errno.ENOENT:
+                    log.error(
+                        "[DDAS-0001-03] AppSec could not read the rule file %s. Reason: file does not exist", self.rules
+                    )
+                else:
+                    # TODO: try to log reasons
+                    log.error("[DDAS-0001-03] AppSec could not read the rule file %s.", self.rules)
+                raise
+            except json.decoder.JSONDecodeError:
+                log.error(
+                    "[DDAS-0001-03] AppSec could not read the rule file %s. Reason: invalid JSON file", self.rules
+                )
+                raise
+            except Exception:
+                # TODO: try to log reasons
+                log.error("[DDAS-0001-03] AppSec could not read the rule file %s.", self.rules)
+                raise
+            try:
+                self._ddwaf = DDWaf(rules)
+            except ValueError:
+                # Partial of DDAS-0005-00
+                log.warning("[DDAS-0005-00] WAF initialization failed")
+                raise
+
+    def on_span_start(self, span):
+        # type: (Span) -> None
+        pass
+
+    def on_span_finish(self, span):
+        # type: (Span) -> None
+        with self._lock:
```

Comment:
The contention on this lock will be huge on multithreaded/async runtimes. Is it required? My understanding of the ddwaf context API is that it is threadsafe.

Reply:
I discussed with Anil, we don't need this for the WAF, only when re-using a context. Let's remove for now